recursive_diff/tests/test_recursive_diff_numpy_types.py

```diff
@@ -0,0 +1,149 @@
+    check(a, b,
+          'object type differs: ndarray<<U...> != ndarray<object>',
+          '[data][0]: 2000-01-01 != 2000-01-01',
+          '[data][1]: 2000-01-02 != 2000-01-02')
+    check(a, c,
+          'object type differs: ndarray<<U...> != ndarray<datetime64>',
+          '[data][0]: 2000-01-01 != 2000-01-01 00:00:00',
+          '[data][1]: 2000-01-02 != 2000-01-02 00:00:00')
+    check(b, c,
+          'object type differs: ndarray<object> != ndarray<datetime64>',
+          '[data][0]: 2000-01-01 != 2000-01-01 00:00:00',
+          '[data][1]: 2000-01-02 != 2000-01-02 00:00:00')
+
+
+def test_numpy_mixed_datetime_types():
+
+    a = np.array(['2000-01-01 12:34:56', '2000-01-02 01:23:45'])
+    b = np.array([datetime.datetime(2000, 1, 1, 12, 34, 56),
+                  datetime.datetime(2000, 1, 2, 1, 23, 45)])
+    c = np.array([np.datetime64('2000-01-01 12:34:56'),
+                  np.datetime64('2000-01-02 01:23:45')])
+    check(a, b,
+          'object type differs: ndarray<<U...> != ndarray<object>',
+          '[data][0]: 2000-01-01 12:34:56 != 2000-01-01 12:34:56',
+          '[data][1]: 2000-01-02 01:23:45 != 2000-01-02 01:23:45')
+    check(a, c,
+          'object type differs: ndarray<<U...> != ndarray<datetime64>',
+          '[data][0]: 2000-01-01 12:34:56 != 2000-01-01 12:34:56',
+          '[data][1]: 2000-01-02 01:23:45 != 2000-01-02 01:23:45')
+    check(b, c,
+          'object type differs: ndarray<object> != ndarray<datetime64>')
+
+
+def test_numpy_mixed_nat_types():
+
+    n1 = np.array(['2000-01-01 12:34:56', 'NaT'])
+    n2 = np.array([np.datetime64('2000-01-01 12:34:56'), np.datetime64('NaT')])
+    check(n1, n2,
+          'object type differs: ndarray<<U...> != ndarray<datetime64>',
+          '[data][0]: 2000-01-01 12:34:56 != 2000-01-01 12:34:56',
+          '[data][1]: NaT != NaT')
+
+
+def test_numpy_same_type_string():
+
+    s1 = np.array(['foo', 'bar', 'bar', 'foo'])
+    s2 = np.array(['foo', 'foo', 'bar', 'bar'])
+
+    check(s1, s2,
+          "[data][1]: bar != foo",
+          "[data][3]: foo != bar")
+
+
+def test_numpy_mixed_string_types():
+
+    so = np.array(['foo', 'bar'], dtype=object)
+    su = np.array(['foo', 'bar'], dtype='U')
+    ss = np.array(['foo', 'bar'], dtype='S')
+
+    check(so, su,
+          "object type differs: ndarray<object> != ndarray<<U...>")
+    check(so, ss,
+          "object type differs: ndarray<object> != ndarray<|S...>",
+          "[data][0]: foo != b'foo'",
+          "[data][1]: bar != b'bar'")
+    check(su, ss,
+          "object type differs: ndarray<<U...> != ndarray<|S...>",
+          "[data][0]: foo != b'foo'",
+          "[data][1]: bar != b'bar'")
+
+
+def test_numpy_same_type_bool():
+
+    b1 = np.array([True, True, False, False], dtype=bool)
+    b2 = np.array([True, False, False, True], dtype=bool)
+
+    check(b1, b2,
+          "[data][1]: True != False",
+          "[data][3]: False != True")
```

Comment:
Yes this test was indeed missing. But please move it to test_recursive_diff.py.
The explicit dtype=bool is unnecessary.